stdlib/bytes.gr

```diff
@@ -0,0 +1,466 @@
+  let size = getSize(ptr)
+  let offset = numberToWasmI32(i)
+  ensureIndexIsInBounds(offset, _INT64_BYTE_SIZE, size)
+  let n = WasmI64.load(ptr + offset, _VALUE_OFFSET)
+  Conv.toInt64(n)
+}
+
+/**
+ * Sets a 64-bit float starting at the given byte index.
+ * @param i: Number - The byte index
+ * @param v: Float64 - The value to set
+ * @param b: Bytes - The Bytes buffer
+ * @returns Void
+ */
+@disableGC
+export let setFloat64 = (i: Number, v: Float64, b: Bytes) => {
+  let (+) = WasmI32.add
+  let ptr = WasmI32.fromGrain(b)
+  let size = getSize(ptr)
+  let offset = numberToWasmI32(i)
+  ensureIndexIsInBounds(offset, _FLOAT64_BYTE_SIZE, size)
+  let value = Conv.fromFloat64(v)
+  WasmF64.store(ptr + offset, value, _VALUE_OFFSET)
+}
+
+/**
+ * Gets a 64-bit float starting at the given byte index.
+ * @param i: Number - The byte index
+ * @param b: Bytes - The Bytes buffer
+ * @returns Float64
+ */
+@disableGC
+export let getFloat64 = (i: Number, b: Bytes) => {
+  let (+) = WasmI32.add
+  let ptr = WasmI32.fromGrain(b)
+  let size = getSize(ptr)
+  let offset = numberToWasmI32(i)
+  ensureIndexIsInBounds(offset, _FLOAT64_BYTE_SIZE, size)
+  let n = WasmF64.load(ptr + offset, _VALUE_OFFSET)
+  Conv.toFloat64(n)
+}
+
+/**
+ * Return a new Bytes buffer that contains the same bytes as the argument.
+ * @param b: Bytes - The Bytes buffer to copy
+ * @returns Bytes
+ */
+@disableGC
+export let copy = (b: Bytes) => {
+  let (+) = WasmI32.add
+  let src = WasmI32.fromGrain(b)
+  let size = getSize(src)
+  let dst = allocateBytes(size)
+  Memory.copy(dst + _VALUE_OFFSET, src + _VALUE_OFFSET, size)
+  WasmI32.toGrain(dst): Bytes
+}
+
+/**
+ * Returns a new Bytes buffer containing a subset of the original Bytes buffer.
+ * @param i: Number - The start position to copy from
+ * @param len: Number - The number of bytes to copy
+ * @param b: Bytes - The Bytes buffer get a subset of bytes from
+ * @returns Bytes
+ */
+@disableGC
+export let sub = (i: Number, len: Number, b: Bytes) => {
+  let (>) = WasmI32.gtS
+  let (+) = WasmI32.add
+  let src = WasmI32.fromGrain(b)
+  let size = getSize(src)
+  let i = numberToWasmI32(i)
+  let len = numberToWasmI32(len)
+  if ((i + len) > size) {
+    throw Exception.InvalidArgument("The given index and length do not specify a valid range of bytes")
+  }
+  let dst = allocateBytes(len)
+  let offset = i
+  Memory.copy(dst + _VALUE_OFFSET, src + _VALUE_OFFSET + i, len)
+  WasmI32.toGrain(dst): Bytes
+}
+
+
+/**
+ * Returns a new Bytes buffer with uninitialized bytes prepended/appended to it.
+ * If the `left` or `right` parameters are negative, bytes will be removed from the corresponding side instead of appended.
+ * @param left: Number - The number of uninitialized bytes to prepend
+ * @param right: Number - The number of uninitialized bytes to append
+ * @param b: Bytes - The Bytes buffer get a subset of bytes from
+ * @returns Bytes
+ */
+@disableGC
+export let extend = (left: Number, right: Number, b: Bytes) => {
```

Comment:
Is this the replacement name for `blit`? I only think of "extend" as "to grow", so the removal seems really weird. I think I'd want to call this `splice`

Reply:
Nah, it's the same as OCaml's `extend`. I like the name `extend` because it's easy to pick out if reading API docs to go for a grow-y kind of action. I think this name is better for how it'll usually be used. That said, I think this could potentially be better as two separate functions, one that only extends and one that only shortens. (I think I'd want to avoid calling them `grow` and `shrink` since that could imply that it's in-place.)

Reply:
I'd be happy with `extend` and `shorten`!

Btw, I noticed the docs say "uninitialized" and I remember that being a big security issue for nodejs because newly created buffers could have old data inside them. They started zero-ing out buffers when created - is that something we need to do?

Reply:
I also think we need to add `grow` and `shrink` because most of the usage of Bytes is mutative based on what I saw in the tests.

Reply:
I think zeroing is also nice from a user's perspective, just has an obvious performance hit. Since there's a bulk memory instruction for it though, I imagine it's not too bad.

Reply:
It's not really possible to implement grow/shrink with Bytes in its current form, since you'd need to be able to reallocate its memory block to a larger size, which isn't possible with our memory manager. I think if you wanted something like Bytes that could grow, you should instead reach for some Buffer type (which we'll also have to implement). They're definitely similar in a lot of ways, but having Bytes with a static size makes it quite a bit more performant.

I guess the behavior is pretty different as well, since a lot of Buffer implementations will automatically grow instead of having a method for it.

Reply:
Yeah I would personally love to skip `grow/shrink` and defer `Buffer` to a separate PR 👀

Reply:
Sounds good! Sorry for my ignorance about the difference 🙈 

Let's not lose my `extend` and `shorten` comment ❤️

Reply:
@phated what should the signatures on those functions be?